# Analyzing CIA Factbook Data Using SQL

## Introduction:
<ul>
<div><p>In this project, we'll work with data from the <a href="https://www.cia.gov/the-world-factbook/" target="_blank">CIA World Factbook</a>, a compendium of statistics about all of the countries on Earth. The Factbook contains demographic information like the following:</p>
<ul>
<li><code>population</code> — the global population.</li>
<li><code>population_growth</code> — the annual population growth rate, as a percentage.</li>
<li><code>area</code> — the total land and water area.</li>
</ul>
<p>In this guided project, we'll use SQL in Jupyter Notebook to analyze data from this database. If you want to work on this project locally, you can download the <a href="https://dsserver-prod-resources-1.s3.amazonaws.com/257/factbook.db" target="_blank">SQLite <code>factbook.db</code> database</a> by clicking the <kbd>Download</kbd> button on the right.</p>
<p>We'll use the following code to connect our Jupyter Notebook to our database file:</p>
</div>
<div class="react-codemirror2 dq-editor dq-mb-4 dq-p-2 dq-bg-gray-50 dark:dq-bg-gray-800"><div class="CodeMirror cm-s-default CodeMirror-wrap"><div style="overflow: hidden; position: relative; width: 3px; height: 0px; top: 47.68px; left: 253.6px;"><textarea autocorrect="off" autocapitalize="off" spellcheck="false" readonly="" tabindex="0" style="position: absolute; bottom: -1em; padding: 0px; width: 1000px; height: 1em; outline: none;"></textarea></div><div class="CodeMirror-vscrollbar" tabindex="-1" cm-not-content="true"><div style="min-width: 1px; height: 0px;"></div></div><div class="CodeMirror-hscrollbar" tabindex="-1" cm-not-content="true"><div style="height: 100%; min-height: 1px; width: 0px;"></div></div><div class="CodeMirror-scrollbar-filler" cm-not-content="true"></div><div class="CodeMirror-gutter-filler" cm-not-content="true"></div><div class="CodeMirror-scroll" tabindex="-1"><div class="CodeMirror-sizer" style="margin-left: 0px; margin-bottom: -14px; border-right-width: 36px; min-height: 79px; padding-right: 0px; padding-bottom: 0px;"><div style="position: relative; top: 0px;"><div class="CodeMirror-lines" role="presentation"><div role="presentation" style="position: relative; outline: none;"><div class="CodeMirror-measure"><pre class="CodeMirror-line-like">x</pre></div><div class="CodeMirror-measure"></div><div style="position: relative; z-index: 1;"></div><div class="CodeMirror-cursors" style="visibility: hidden;"><div class="CodeMirror-cursor" style="left: 253.6px; top: 39.68px; height: 19.84px;">&nbsp;</div></div><div class="CodeMirror-code" role="presentation"><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">%%capture</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">%load_ext sql</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">%sql sqlite:///factbook.db</span></pre></div></div></div></div></div><div style="position: absolute; height: 36px; width: 1px; border-bottom: 0px solid transparent; top: 79px;"></div><div class="CodeMirror-gutters" style="display: none; height: 115px;"></div>

### Things we will do:
<div class="dq-px-8"><div id="body" class="dq-lesson-markup dq-lesson-learn-instructions dq-max-w-screen-xl"><ol>
<li>Copy the text above into a code cell in the Jupyter Notebook on the right.</li>
<li>Using the menu bar or keyboard shortcut, run the cell.</li>


In [1]:
%%capture
%load_ext sql
%sql sqlite:///factbook.db

## Overview of the Data:
<ul>
<div id="body" class="dq-lesson-markup dq-max-w-screen-xl"><div><p>Throughout this course, we introduced the name of the table and what the table looked like in the <code>jobs.db</code> database. We can actually query the database to get this information directly:</p>
</div><div class="react-codemirror2 dq-editor dq-mb-4 dq-p-2 dq-bg-gray-50 dark:dq-bg-gray-800"><div class="CodeMirror cm-s-default CodeMirror-wrap"><div style="overflow: hidden; position: relative; width: 3px; height: 0px; top: 47.68px; left: 196px;"><textarea autocorrect="off" autocapitalize="off" spellcheck="false" readonly="" tabindex="0" style="position: absolute; bottom: -1em; padding: 0px; width: 1000px; height: 1em; outline: none;"></textarea></div><div class="CodeMirror-vscrollbar" tabindex="-1" cm-not-content="true"><div style="min-width: 1px; height: 0px;"></div></div><div class="CodeMirror-hscrollbar" tabindex="-1" cm-not-content="true"><div style="height: 100%; min-height: 1px; width: 0px;"></div></div><div class="CodeMirror-scrollbar-filler" cm-not-content="true"></div><div class="CodeMirror-gutter-filler" cm-not-content="true"></div><div class="CodeMirror-scroll" tabindex="-1"><div class="CodeMirror-sizer" style="margin-left: 0px; margin-bottom: -14px; border-right-width: 36px; min-height: 79px; padding-right: 0px; padding-bottom: 0px;"><div style="position: relative; top: 0px;"><div class="CodeMirror-lines" role="presentation"><div role="presentation" style="position: relative; outline: none;"><div class="CodeMirror-measure"><pre class="CodeMirror-line-like"><span>xxxxxxxxxx</span></pre></div><div class="CodeMirror-measure"></div><div style="position: relative; z-index: 1;"></div><div class="CodeMirror-cursors" style="visibility: hidden;"><div class="CodeMirror-cursor" style="left: 196px; top: 39.68px; height: 19.84px;">&nbsp;</div></div><div class="CodeMirror-code" role="presentation"><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-keyword">SELECT</span> <span class="cm-operator">*</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp;<span class="cm-keyword">FROM</span> sqlite_master</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> <span class="cm-keyword">WHERE</span> type<span class="cm-operator">=</span><span class="cm-string">'table'</span><span class="cm-punctuation">;</span></span></pre></div></div></div></div></div><div style="position: absolute; height: 36px; width: 1px; border-bottom: 0px solid transparent; top: 79px;"></div><div class="CodeMirror-gutters" style="display: none; height: 115px;"></div></div></div></div><div>
<p>This will return the following output:</p>
<p></p><center>
<table>
<tbody><tr>
<th></th>
<th>type</th>
<th>name</th>
<th>tbl_name</th>
<th>rootpage</th>
<th>sql</th>
</tr>
<tr>
<td>0</td>
<td>table</td>
<td>facts</td>
<td>facts</td>
<td>2</td>
<td>CREATE TABLE "facts" ("id" INTEGER PRIMARY KEY...</td>
</tr>
<tr>
<td>1</td>
<td>table</td>
<td>sqlite_sequence</td>
<td>sqlite_sequence</td>
<td>3</td>
<td>CREATE TABLE sqlite_sequence(name,seq)</td>
</tr>
</tbody></table>
</center><p></p>
<p>To run SQL queries in this project, we add <code>%%sql</code> on its own line to the start of our query.  So to execute the query above, we'll use the following code:</p>
</div><div class="react-codemirror2 dq-editor dq-mb-4 dq-p-2 dq-bg-gray-50 dark:dq-bg-gray-800"><div class="CodeMirror cm-s-default CodeMirror-wrap"><div style="overflow: hidden; position: relative; width: 3px; height: 0px; top: 67.52px; left: 196px;"><textarea autocorrect="off" autocapitalize="off" spellcheck="false" readonly="" tabindex="0" style="position: absolute; bottom: -1em; padding: 0px; width: 1000px; height: 1em; outline: none;"></textarea></div><div class="CodeMirror-vscrollbar" tabindex="-1" cm-not-content="true"><div style="min-width: 1px; height: 0px;"></div></div><div class="CodeMirror-hscrollbar" tabindex="-1" cm-not-content="true"><div style="height: 100%; min-height: 1px; width: 0px;"></div></div><div class="CodeMirror-scrollbar-filler" cm-not-content="true"></div><div class="CodeMirror-gutter-filler" cm-not-content="true"></div><div class="CodeMirror-scroll" tabindex="-1"><div class="CodeMirror-sizer" style="margin-left: 0px; margin-bottom: -13px; border-right-width: 37px; min-height: 99px; padding-right: 0px; padding-bottom: 0px;"><div style="position: relative; top: 0px;"><div class="CodeMirror-lines" role="presentation"><div role="presentation" style="position: relative; outline: none;"><div class="CodeMirror-measure"><pre class="CodeMirror-line-like"><span>xxxxxxxxxx</span></pre></div><div class="CodeMirror-measure"></div><div style="position: relative; z-index: 1;"></div><div class="CodeMirror-cursors"><div class="CodeMirror-cursor" style="left: 196px; top: 59.52px; height: 19.84px;">&nbsp;</div></div><div class="CodeMirror-code" role="presentation"><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-operator">%%</span>sql</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-keyword">SELECT</span> <span class="cm-operator">*</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp;<span class="cm-keyword">FROM</span> sqlite_master</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> <span class="cm-keyword">WHERE</span> type<span class="cm-operator">=</span><span class="cm-string">'table'</span><span class="cm-punctuation">;</span></span></pre></div></div></div></div></div><div style="position: absolute; height: 37px; width: 1px; border-bottom: 0px solid transparent; top: 99px;"></div><div class="CodeMirror-gutters" style="display: none; height: 136px;"></div></div></div></div><div>
<p>Let's get started writing queries!</p></div>
<ol>
<li>We'll write a query to return information on the tables in the database.</li>
<li>In a different code cell, We'll write and run another query that returns the first five rows of the <code>facts</code> table in the database.</li>
</ol>

In [2]:
%%sql

SELECT *

  FROM sqlite_master

 WHERE type='table';

 * sqlite:///factbook.db
Done.


[('table', 'sqlite_sequence', 'sqlite_sequence', 3, 'CREATE TABLE sqlite_sequence(name,seq)'),
 ('table', 'facts', 'facts', 47, 'CREATE TABLE "facts" ("id" INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL, "code" varchar(255) NOT NULL, "name" varchar(255) NOT NULL, "area" integer, "a ... (4 characters truncated) ... land" integer, "area_water" integer, "population" integer, "population_growth" float, "birth_rate" float, "death_rate" float, "migration_rate" float)')]

In [3]:
%%sql

SELECT *
  FROM facts
 LIMIT 5;

 * sqlite:///factbook.db
Done.


[(1, 'af', 'Afghanistan', 652230, 652230, 0, 32564342, 2.32, 38.57, 13.89, 1.51),
 (2, 'al', 'Albania', 28748, 27398, 1350, 3029278, 0.3, 12.92, 6.58, 3.3),
 (3, 'ag', 'Algeria', 2381741, 2381741, 0, 39542166, 1.84, 23.67, 4.31, 0.92),
 (4, 'an', 'Andorra', 468, 468, 0, 85580, 0.12, 8.13, 6.96, 0.0),
 (5, 'ao', 'Angola', 1246700, 1246700, 0, 19625353, 2.78, 38.78, 11.49, 0.46)]

## Summary Statistics:
<ul>
<div id="body" class="dq-lesson-markup dq-max-w-screen-xl"><div><p>Executing our previous query, we see the first few rows of our <code>facts</code> table look like this:</p>
<p></p><center>
<table>
<tbody>
<tr>
<th>id</th>
<th>code</th>
<th>name</th>
<th>area</th>
<th>area_land</th>
<th>area_water</th>
<th>population</th>
<th>population_growth</th>
<th>birth_rate</th>
<th>death_rate</th>
<th>migration_rate</th>
</tr>
<tr>
<td>1</td>
<td>af</td>
<td>Afghanistan</td>
<td>652230</td>
<td>652230</td>
<td>0</td>
<td>32564342</td>
<td>2.32</td>
<td>38.57</td>
<td>13.89</td>
<td>1.51</td>
</tr>
<tr>
<td>2</td>
<td>al</td>
<td>Albania</td>
<td>28748</td>
<td>27398</td>
<td>1350</td>
<td>3029278</td>
<td>0.3</td>
<td>12.92</td>
<td>6.58</td>
<td>3.3</td>
</tr>
<tr>
<td>3</td>
<td>ag</td>
<td>Algeria</td>
<td>2381741</td>
<td>2381741</td>
<td>0</td>
<td>39542166</td>
<td>1.84</td>
<td>23.67</td>
<td>4.31</td>
<td>0.92</td>
</tr>
<tr>
<td>4</td>
<td>an</td>
<td>Andorra</td>
<td>468</td>
<td>468</td>
<td>0</td>
<td>85580</td>
<td>0.12</td>
<td>8.13</td>
<td>6.96</td>
<td>0.0</td>
</tr>
<tr>
<td>5</td>
<td>ao</td>
<td>Angola</td>
<td>1246700</td>
<td>1246700</td>
<td>0</td>
<td>19625353</td>
<td>2.78</td>
<td>38.78</td>
<td>11.49</td>
<td>0.46</td>
</tr>
</tbody>
</table>
</center><p></p>
<p>Here are the descriptions for some of the columns:</p>
<ul>
<li><code>name</code> — the name of the country.</li>
<li><code>area</code>— the country's total area (both land and water).</li>
<li><code>area_land</code> — the country's land area in <a href="https://www.cia.gov/library/publications/the-world-factbook/rankorder/2147rank.html" target="_blank">square kilometers</a>.</li>
<li><code>area_water</code> — the country's waterarea in square kilometers.</li>
<li><code>population</code> — the country's population.</li>
<li><code>population_growth</code>— the country's population growth as a percentage.</li>
<li><code>birth_rate</code> — the country's birth rate, or the number of births per year per 1,000 people.</li>
<li><code>death_rate</code> — the country's death rate, or the number of death per year per 1,000 people.</li>
</ul>
<p>Let's start by calculating some summary statistics and look for any outlier countries.</p></div></div>

Next we will do:
<div id="body" class="dq-lesson-markup dq-lesson-learn-instructions dq-max-w-screen-xl"><ol>
<li>
<p>Write a single query that returns the following:</p>
<ul>
<li>Minimum population</li>
<li>Maximum population</li>
<li>Minimum population growth </li>
<li>Maximum population growth</li>
</ul>
</li>
</ol></div>

In [4]:
%%sql
SELECT MIN(population), MAX(population), MIN(population_growth), MAX(population_growth)
  FROM facts;

 * sqlite:///factbook.db
Done.


[(0, 7256490011, 0.0, 4.02)]

## Exploring Outliers:
<ul>
<div><p>We see a few interesting things in the summary statistics on the previous screen:</p>
<ul>
<li>There's a country with a population of <code>0</code></li>
<li>There's a country with a population of <code>7256490011</code> (or more than 7.2 billion people) </li>
</ul>
<p>Let's use subqueries to zoom in on just these countries <em>without</em> using the specific values.</p></div></div>

We'll:
<div id="body" class="dq-lesson-markup dq-lesson-learn-instructions dq-max-w-screen-xl"><ol>
<li>Write a query that returns the countries with the minimum population.</li>
<li>Write a query that returns the countries with the maximum population.</li>
</ol></div>

In [5]:
%%sql
SELECT name, population
  FROM facts
 GROUP BY population
HAVING population == (SELECT MIN(population)
                        FROM facts);

 * sqlite:///factbook.db
Done.


[('Antarctica', 0)]

In [6]:
%%sql
SELECT name, population
  FROM facts
 GROUP BY population
HAVING population == (SELECT MAX(population)
                        FROM facts);

 * sqlite:///factbook.db
Done.


[('World', 7256490011)]

In [7]:
%%sql
SELECT name, population AS 'population ascending'
  FROM facts
 GROUP BY population
 ORDER BY 'population ascending' ASC
 LIMIT 6;


 * sqlite:///factbook.db
Done.


[('Ashmore and Cartier Islands', None),
 ('Antarctica', 0),
 ('Pitcairn Islands', 48),
 ('Cocos (Keeling) Islands', 596),
 ('Holy See (Vatican City)', 842),
 ('Niue', 1190)]

In [8]:
%%sql
SELECT name, population AS 'population descending'
  FROM facts
 GROUP BY population
 ORDER BY 'population descending' DESC 
 LIMIT 6;

 * sqlite:///factbook.db
Done.


[('World', 7256490011),
 ('China', 1367485388),
 ('India', 1251695584),
 ('European Union', 513949445),
 ('United States', 321368864),
 ('Indonesia', 255993674)]

## Exploring Average Population and Area:
<ul>
<div id="body" class="dq-lesson-markup dq-max-w-screen-xl"><div><p>It seems like the table contains a row for the whole world, which explains the population of over 7.2 billion. It also seems like the table contains a row for Antarctica, which explains the population of 0. This seems to match the CIA Factbook <a href="https://www.cia.gov/library/publications/the-world-factbook/geos/ay.html" target="_blank">page for Antarctica</a>:</p>
<p><img src="https://s3.amazonaws.com/dq-content/257/fb_antarctica.png"></p>
<p>Now that we know this, we should recalculate the summary statistics we calculated earlier — this time excluding the row for the whole world.</p></div></div>
  
Next we'll:
<div id="body" class="dq-lesson-markup dq-lesson-learn-instructions dq-max-w-screen-xl"><ol>
<li>
<p>Recompute the summary statistics you found earlier while excluding the row for the whole world. Include the following:</p>
<ul>
<li>Minimum population</li>
<li>Maximum population</li>
<li>Minimum population growth </li>
<li>Maximum population growth</li>
</ul>
</li>
<li>
<p>In a different code cell, calculate the average value for the following columns:</p>
<ul>
<li><code>population</code></li>
<li><code>area</code></li>
</ul>
</li>
</ol></div>

In [9]:
%%sql

SELECT name, MIN(population), MAX(population), MIN(population_growth), MAX(population_growth)
  FROM facts
 WHERE name != 'World';

 * sqlite:///factbook.db
Done.


[('South Sudan', 0, 1367485388, 0.0, 4.02)]

In [10]:
%%sql
SELECT AVG(population), AVG(area)
  FROM facts;

 * sqlite:///factbook.db
Done.


[(62094928.32231405, 555093.546184739)]

## Finding Densely Populated Countries:
<ul>
<div class="dq-px-8"><div id="body" class="dq-lesson-markup dq-max-w-screen-xl"><div><p>To finish, we'll build on the query we wrote for the previous screen to find countries that are densely populated.  We'll identify countries that have the following:</p>
<ul>
<li>Above-average values for population.</li>
<li>Below-average values for area.</li>
</ul>
<p>It's possible to write this query by copying the values we calculated on the previous screen, but also to do this without hard-coding them!</p></div></div></div>
    
Next we'll:
<div id="body" class="dq-lesson-markup dq-lesson-learn-instructions dq-max-w-screen-xl"><ol>
<li>Write a query that finds all countries meeting both of the following criteria:<ul>
<li>The <code>population</code> is above average.</li>
<li>The <code>area</code> is below average.</li>
</ul>
</li>
</ol></div>

In [11]:
%%sql

SELECT *
  FROM facts
 WHERE population > (SELECT AVG(population)
                       FROM facts
                      WHERE name != 'World') -- # We need this exclusion for a proper aggregation.
   AND area < (SELECT AVG(area)
              FROM facts
             WHERE name != 'World'); -- # We need this exclusion for a proper aggregation.

 * sqlite:///factbook.db
Done.


[(14, 'bg', 'Bangladesh', 148460, 130170, 18290, 168957745, 1.6, 21.14, 5.61, 0.46),
 (65, 'gm', 'Germany', 357022, 348672, 8350, 80854408, 0.17, 8.47, 11.42, 1.24),
 (80, 'iz', 'Iraq', 438317, 437367, 950, 37056169, 2.93, 31.45, 3.77, 1.62),
 (83, 'it', 'Italy', 301340, 294140, 7200, 61855120, 0.27, 8.74, 10.19, 4.1),
 (85, 'ja', 'Japan', 377915, 364485, 13430, 126919659, 0.16, 7.93, 9.51, 0.0),
 (91, 'ks', 'Korea, South', 99720, 96920, 2800, 49115196, 0.14, 8.19, 6.75, 0.0),
 (120, 'mo', 'Morocco', 446550, 446300, 250, 33322699, 1.0, 18.2, 4.81, 3.36),
 (138, 'rp', 'Philippines', 300000, 298170, 1830, 100998376, 1.61, 24.27, 6.11, 2.09),
 (139, 'pl', 'Poland', 312685, 304255, 8430, 38562189, 0.09, 9.74, 10.19, 0.46),
 (163, 'sp', 'Spain', 505370, 498980, 6390, 48146134, 0.89, 9.64, 9.04, 8.31),
 (173, 'th', 'Thailand', 513120, 510890, 2230, 67976405, 0.34, 11.19, 7.8, 0.0),
 (182, 'ug', 'Uganda', 241038, 197100, 43938, 37101745, 3.24, 43.79, 10.69, 0.74),
 (185, 'uk', 'United Kingdom', 243610, 241930, 1680, 64088222, 0.54, 12.17, 9.35, 2.54),
 (192, 'vm', 'Vietnam', 331210, 310070, 21140, 94348835, 0.97, 15.96, 5.93, 0.3)]

Let's create two versions of this chart:
- one where we can see the smallest area_land (in ascending order).
- another, where the population is the greatest (in descending order).

In [12]:
%%sql

SELECT *
  FROM facts
 WHERE population > (SELECT AVG(population)
                       FROM facts
                      WHERE name != 'World') -- # We need this exclusion for a proper aggregation.
   AND area < (SELECT AVG(area)
              FROM facts
             WHERE name != 'World') -- # We need this exclusion for a proper aggregation.
 ORDER BY area ASC;

 * sqlite:///factbook.db
Done.


[(91, 'ks', 'Korea, South', 99720, 96920, 2800, 49115196, 0.14, 8.19, 6.75, 0.0),
 (14, 'bg', 'Bangladesh', 148460, 130170, 18290, 168957745, 1.6, 21.14, 5.61, 0.46),
 (182, 'ug', 'Uganda', 241038, 197100, 43938, 37101745, 3.24, 43.79, 10.69, 0.74),
 (185, 'uk', 'United Kingdom', 243610, 241930, 1680, 64088222, 0.54, 12.17, 9.35, 2.54),
 (138, 'rp', 'Philippines', 300000, 298170, 1830, 100998376, 1.61, 24.27, 6.11, 2.09),
 (83, 'it', 'Italy', 301340, 294140, 7200, 61855120, 0.27, 8.74, 10.19, 4.1),
 (139, 'pl', 'Poland', 312685, 304255, 8430, 38562189, 0.09, 9.74, 10.19, 0.46),
 (192, 'vm', 'Vietnam', 331210, 310070, 21140, 94348835, 0.97, 15.96, 5.93, 0.3),
 (65, 'gm', 'Germany', 357022, 348672, 8350, 80854408, 0.17, 8.47, 11.42, 1.24),
 (85, 'ja', 'Japan', 377915, 364485, 13430, 126919659, 0.16, 7.93, 9.51, 0.0),
 (80, 'iz', 'Iraq', 438317, 437367, 950, 37056169, 2.93, 31.45, 3.77, 1.62),
 (120, 'mo', 'Morocco', 446550, 446300, 250, 33322699, 1.0, 18.2, 4.81, 3.36),
 (163, 'sp', 'Spain', 505370, 498980, 6390, 48146134, 0.89, 9.64, 9.04, 8.31),
 (173, 'th', 'Thailand', 513120, 510890, 2230, 67976405, 0.34, 11.19, 7.8, 0.0)]

In [13]:
%%sql

SELECT *
  FROM facts
 WHERE population > (SELECT AVG(population)
                       FROM facts
                      WHERE name != 'World') -- # We need this exclusion for a proper aggregation.
   AND area < (SELECT AVG(area)
              FROM facts
             WHERE name != 'World') -- # We need this exclusion for a proper aggregation.
 ORDER BY population DESC;

 * sqlite:///factbook.db
Done.


[(14, 'bg', 'Bangladesh', 148460, 130170, 18290, 168957745, 1.6, 21.14, 5.61, 0.46),
 (85, 'ja', 'Japan', 377915, 364485, 13430, 126919659, 0.16, 7.93, 9.51, 0.0),
 (138, 'rp', 'Philippines', 300000, 298170, 1830, 100998376, 1.61, 24.27, 6.11, 2.09),
 (192, 'vm', 'Vietnam', 331210, 310070, 21140, 94348835, 0.97, 15.96, 5.93, 0.3),
 (65, 'gm', 'Germany', 357022, 348672, 8350, 80854408, 0.17, 8.47, 11.42, 1.24),
 (173, 'th', 'Thailand', 513120, 510890, 2230, 67976405, 0.34, 11.19, 7.8, 0.0),
 (185, 'uk', 'United Kingdom', 243610, 241930, 1680, 64088222, 0.54, 12.17, 9.35, 2.54),
 (83, 'it', 'Italy', 301340, 294140, 7200, 61855120, 0.27, 8.74, 10.19, 4.1),
 (91, 'ks', 'Korea, South', 99720, 96920, 2800, 49115196, 0.14, 8.19, 6.75, 0.0),
 (163, 'sp', 'Spain', 505370, 498980, 6390, 48146134, 0.89, 9.64, 9.04, 8.31),
 (139, 'pl', 'Poland', 312685, 304255, 8430, 38562189, 0.09, 9.74, 10.19, 0.46),
 (182, 'ug', 'Uganda', 241038, 197100, 43938, 37101745, 3.24, 43.79, 10.69, 0.74),
 (80, 'iz', 'Iraq', 438317, 437367, 950, 37056169, 2.93, 31.45, 3.77, 1.62),
 (120, 'mo', 'Morocco', 446550, 446300, 250, 33322699, 1.0, 18.2, 4.81, 3.36)]

## Conclusion:

We encountered the two most extreme countries in population and area way are:
- `Bangladesh`
- `Korea, South`

What's interesting is both are in Asia. Another interesting fact is that [Korea, South](https://en.wikipedia.org/wiki/South_Korea) is a very developed country, where [Bangladesh](https://en.wikipedia.org/wiki/Bangladesh) is in the opposite. 

From that, we can conclude, that a small area isn't a factor that can give us any insight into development.
On the other hand, extreme population growth can be a factor of destitution, poverty.

In [23]:
%%sql

SELECT * FROM facts
 INNER JOIN cities ON cities.facts_id = facts.id
 LIMIT 10; 


# SELECT *
#   FROM cities AS c, FROM facts AS f
#  INNER JOIN cities AS c ON c.* = f.name;

 * sqlite:///factbook.db
(sqlite3.OperationalError) no such table: cities
[SQL: SELECT * FROM facts
 INNER JOIN cities ON cities.facts_id = facts.id
 LIMIT 10;]
(Background on this error at: http://sqlalche.me/e/e3q8)
